### PR TITLE
RavenDB-17691: Simplified the code and added extra checks

### PR DESCRIPTION
**Pull request by redknightlois**

### CODE REVIEW

**Comment by redknightlois** [comment]
I couldnt really find anything at that location that could cause an AVE. Nevertheless, thinking that maybe we are triggering a codegen issue, I modified the code in order to try to find some signal in the noise. BTW, cannot reproduce the issue on my current setup. 

cc @arekpalinski

File: src/Corax/Queries/BinaryMatch.Boolean.cs [hunk 1]
```diff
@@ -82,7 +83,9 @@ static int FillFunc(ref BinaryMatch<TInner, TOuter> match, Span<long> matches)
                 // need to be ready to put both outputs to the matches
                 int length = matches.Length / 2;
                 Span<long> innerMatches = longBuffer.Slice(0, length);
+                Debug.Assert(innerMatches.Length == matches.Length / 2);
```
Comment (ayende) [inline src/Corax/Queries/BinaryMatch.Boolean.cs]: Let's check that `length is > 0`

File: src/Corax/Queries/BinaryMatch.Boolean.cs [hunk 2]
```diff
@@ -81,8 +82,16 @@ static int FillFunc(ref BinaryMatch<TInner, TOuter> match, Span<long> matches)
...
                 // need to be ready to put both outputs to the matches
                 int length = matches.Length / 2;
+                if (length < 16)
+                {
+                    // In case we get smaller matches we still acquire a cache line. 
+                    length = 16;
+                }                
+
                 Span<long> innerMatches = longBuffer.Slice(0, length);
+                Debug.Assert(innerMatches.Length == matches.Length / 2);
```
Comment (ayende) [inline src/Corax/Queries/BinaryMatch.Boolean.cs]: That will fail if matches.Length is < 32

Comment (redknightlois) [reply src/Corax/Queries/BinaryMatch.Boolean.cs]: Got rid of all that, it was flawed by design. The only real edge case is `matches.Length == 1` which requires a different way to fill.

**Comment by redknightlois** [comment]
Need Jenkins to run the tests.

File: src/Corax/Queries/TermMatch.cs [hunk 3]
```diff
@@ -273,77 +277,91 @@ static int AndWithVectorizedFunc(ref TermMatch term, Span<long> matches)
...
                     }
 
                     inputPtr = read < matches.Length ? largerPtr : smallerPtr;
 
                     Debug.Assert(inputPtr >= dstPtr);
+                    Debug.Assert((read >= matches.Length ? largerPtr : smallerPtr) - blockStartPtr <= BlockSize);
                 }
 
-                return (int)((ulong)dstPtr - (ulong)Unsafe.AsPointer(ref MemoryMarshal.GetReference(matches))) / sizeof(ulong);
+                return (int)((ulong)dstPtr - (ulong)inputStartPtr) / sizeof(ulong);
```
Comment (maciejaszyk) [inline src/Corax/Queries/TermMatch.cs:364]: Why are you dividing by `sizeof(ulong)`?  `inputStartPtr` and `dstPtr` are `long*`, not `byte*`.

Comment (redknightlois) [reply src/Corax/Queries/TermMatch.cs:364]: This is a trick to improve codegen (tighter). The JIT is quite bad at generating differences of pointers. What this will do is:

- Take the byte addressable location in memory for `dstPtr` and `inputStartPtr` as a `ulong`
- Calculate the amount of bytes in difference between the 2 locations
- Divide by the size of the element (`ulong` in this case). 
- Cast back to int (mask out the higher part of the register).

The reason why? Is because that division now will be picked by the JIT and do a shift instead of using the multiplication ports.
